server/aesdsocket.c

```diff
@@ -81,27 +117,246 @@ void sig_handler(int sig)
+    if(0 == characters_written)
+    {
+        perror("strftime()");
+        error_handler();
+        exit(ERROR_CODE);
+    }
+
+    //Use mutex lock when using the shared resource i.e., when writing to /var/tmp/aesdsocketdata. 
+    if(-1 == pthread_mutex_lock(&mutex))
+    {
+        perror("pthread_mutex_lock()");
+        error_handler();
+        exit(ERROR_CODE);
+    }
+    if(-1 == write(file_des->fd,t_buf,characters_written))
+    {
+        perror("write()");
+        error_handler();
+        exit(ERROR_CODE);
+    }
+    //Use mutex unlock after using the shared resource i.e., after writing to /var/tmp/aesdsocketdata. 
+    if(-1 == pthread_mutex_unlock(&mutex))
+    {
+        perror("pthread_mutex_unlock()");
+        error_handler();
+        exit(ERROR_CODE);
+    }
+}
+
+void* thread_routine(void *arg)
+{
+    int present_location = 0;
+    int size = 0;
+    int bytes_read = 0;
+    int extra_alloc = 1;
+    thread_entries_t *routine_values = (thread_entries_t*)arg;
+    routine_values->d_buf = calloc(initial_alloc_size,sizeof(char));
+    while((bytes_read = recv(routine_values->client_sock,routine_values->d_buf+present_location,initial_alloc_size,0)) > 0)
+    {
+
+        if(-1 == bytes_read)
+        {
+            perror("resv()");
+            error_handler();
+            exit(ERROR_CODE);          
+        } 
+        //If new line is not received, need to keep incrementing the present location to avoid overwritting the previous data received
+		present_location+=bytes_read;           
+        /*
+        * Reference to strchr function : https://man7.org/linux/man-pages/man3/strchr.3.html
+        */
+        if(strchr(routine_values->d_buf ,'\n') != NULL)
+        {
+            break;
+        }
+				
+        extra_alloc++;
+        routine_values->d_buf = (char*)realloc(routine_values->d_buf,(extra_alloc*initial_alloc_size)*sizeof(char));
+		if(NULL == routine_values->d_buf)
+		{
+			syslog(LOG_ERR,"Error: realloc()");
+            //free(routine_values->d_buf);
+            error_handler();
+            exit(ERROR_CODE);
+		}
+    }
+
+    //Use mutex lock when using the shared resource i.e., when writing to /var/tmp/aesdsocketdata. 
+    if(-1 == pthread_mutex_lock(routine_values->my_mutex))
+    {
+        perror("pthread_mutex_lock()");
+        error_handler();
+        exit(ERROR_CODE);
+    }
+
+    if(-1 == write(routine_values->fd,routine_values->d_buf,present_location))
+    {
+        perror("write()");
+		error_handler();
+		exit(ERROR_CODE);    
+    }
+
+    if(-1 == pthread_mutex_unlock(routine_values->my_mutex))
+    {
+        perror("pthread_mutex_unlock()");
+        error_handler();
+        exit(ERROR_CODE);
+    }    
+
+    /*
+    * Get the lenght of the file and set the cursor to the start of the file
+    */
+    size = lseek(routine_values->fd,0,SEEK_END);
+    lseek(routine_values->fd,0,SEEK_SET);
+    routine_values->send_to_client_buf = calloc(size,sizeof(char));
```

Comment:
@Rishikesh38 what does this assume about the total size of all packets sent relative to the size of available RAM?